opengovsg/askgovsg
# feat(search): add createPost and updatePost for search entries

## Problem

No infrastructure to add and update a search entry to synchronise opensearch index with database when posts are added to the database.

Working towards #687

## Solution

Add sync service to include createPost and updatePost, which is called by postService to synchronise opensearch's index with the database. Includes relevant unit tests as well.

## Tests

Setup:

1. Login as an administrator.
2. Open the network panel so as to observe network calls.

### Test for success use cases

3. Create a post. A POST request to `/posts` should be observed in the network panel as usual, as seen in the screenshot:

    <img width="726" alt="Screenshot 2021-12-13 at 2 08 42 PM" src="https://user-images.githubusercontent.com/37061143/145761415-100c99a5-f5bc-4ffb-bd69-be371b73002d.png">

    This means that the POST request to add the post was successful. Payload included in the request should be something like the one below. Take note of the values in the `title` and `description` fields:

    <img width="725" alt="Screenshot 2021-12-13 at 2 08 50 PM" src="https://user-images.githubusercontent.com/37061143/145761433-1e9f93a2-bfb4-4ca1-b22c-8d4373652bc6.png">

    The doc `_id`, which is set to the post id, can be seen in the response, which is **44** in this example as seen below.

    <img width="726" alt="Screenshot 2021-12-13 at 2 09 01 PM" src="https://user-images.githubusercontent.com/37061143/145761439-51a89e0b-e9fd-4d4d-a255-4b574179d765.png">

    This value should match the post id in the URL.

4. Check that the post has been indexed on opensearch with a GET request `curl -XGET 'https://localhost:9200/search_entries/_doc/44' --insecure -u 'admin:admin'`. `_version` should be 1 as the post is freshly added.

    <img width="1103" alt="Screenshot 2021-12-13 at 2 09 36 PM" src="https://user-images.githubusercontent.com/37061143/145761952-4b4c7e03-8c46-463c-ae36-8dc528c784e5.png">

5. Edit the post. A PUT request to `/post/<post_id>` should be observed in the network panel as usual, as seen in the screenshot.

    <img width="726" alt="Screenshot 2021-12-13 at 2 19 12 PM" src="https://user-images.githubusercontent.com/37061143/145762321-c9c0cee6-30e1-45ae-b8ec-08dd6ad7ffa1.png">

    This means that the PUT request to add the post was successful. Payload included in the request should be something like the one below. Take note of the updated values in the `title` and `description` fields:

    <img width="725" alt="Screenshot 2021-12-13 at 2 19 53 PM" src="https://user-images.githubusercontent.com/37061143/145762309-9e942ad9-19da-4b49-bd68-74c5960fad9a.png">

    The response should say that the answer has been updated.

    <img width="726" alt="Screenshot 2021-12-13 at 2 19 26 PM" src="https://user-images.githubusercontent.com/37061143/145762316-e4c0ad52-a1e7-4421-a23c-b3a8ebfb4321.png">

6. Again, check that the post has been updated on opensearch with a GET request `curl -XGET 'https://localhost:9200/search_entries/_doc/44' --insecure -u 'admin:admin'`. `_version` should be 2 as the entry has been changed once after creation.

    <img width="1103" alt="Screenshot 2021-12-13 at 2 20 07 PM" src="https://user-images.githubusercontent.com/37061143/145762304-3a739e1f-39f5-4efa-a473-c032cfab878a.png">

### Test for failure use cases

To simulate errors being thrown by opensearch's API, simply change code in the sync service (`server/src/modules/search/sync/sync.service.ts`) such that it will fail. For instance, `this.client.create` can be changed to `this.client.update` and vice versa. Rebuilding the file after these code changes is not required as the server is hot-reloaded.

7. Create a post. The post creation should fail with status code `500 Internal Server Error`.

    <img width="725" alt="Screenshot 2021-12-13 at 2 33 04 PM" src="https://user-images.githubusercontent.com/37061143/145763690-ea946665-5d14-4faa-ad22-35a1d4865145.png">

    Post should not show up on the database or opensearch index as well.

8. Update a post. Similar to the previous point, the request should fail with `500 Internal Server Error` and fields which were edited should not be updated on both the database and opensearch index.
